### PR TITLE
add agent sidecar flame for implementation

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <i.can.feel.gravity@gmail.com>

### Description:
We decided to implement blob storage backup functionality.
so, I added a prototype of agent sidecar flame, @rinx please implement backup logic after this PR merged.

### Environment:

- Golang Version: 1.14.3
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.11.5

### Types of changes:

- [ ] Bug fix [type/bug]
- [x] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/agent/sidecar/handler/grpc/option.go [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+
+type Option func(*server)
+
+var (
+	defaultOpts = []Option{}
+)
+
+func WithStorageObserver(so service.StorageObserver) Option {
+	return func(s *server) {
+		if i != nil {
```
Comment (rinx) [inline pkg/agent/sidecar/handler/grpc/option.go:32]: The build is failing caused by this `i`. Please fix it :wink:
https://github.com/vdaas/vald/pull/404/checks?check_run_id=706185779

File: internal/file/watch/option.go [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+		if w.dirs != nil {
+			w.dirs = append(w.dirs, dirs...)
+		} else {
+			w.dirs = dirs
+		}
+		return nil
+	}
+}
+
+func WithDir(dir string) Option {
```
Comment (rinx) [inline internal/file/watch/option.go:54]: This function is useless because there's `WithDirs` func.

File: internal/file/watch/watch.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+		)
+		handleErr := func(ctx context.Context, err error) {
+			log.Error(err)
+			select {
+			case <-ctx.Done():
+			case ech <- err:
+			}
+		}
+		for {
+			ok = true
```
Comment (rinx) [inline internal/file/watch/watch.go:102]: this line should be redundant. right?

File: internal/file/watch/watch.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+						err = w.onDelete(ctx, event.Name)
+					case event.Op&fsnotify.Rename == fsnotify.Rename && w.onRename != nil:
+						log.Debug("Renamed file: ", event.Name)
+						err = w.onRename(ctx, event.Name)
+					case event.Op&fsnotify.Chmod == fsnotify.Chmod && w.onChmod != nil:
+						log.Debug("File changed permission: ", event.Name)
+						err = w.onChmod(ctx, event.Name)
+					}
+				}
+			}
```
Comment (rinx) [inline internal/file/watch/watch.go:135]: badge
```suggestion
			case err, ok = <-w.w.Errors:
			}
```
and remove the goroutine started from L145.
it's better to reduce the number of useless goroutines.

File: internal/config/sidecar.go [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+// Package config providers configuration type and load configuration logic
+package config
+
+type Sidecar struct {
```
Comment (rinx) [inline internal/config/sidecar.go:20]: badge
Currently, sidecar component is only for agent. however, in the future, other sidecar components may be required.
So it should be named as "AgentSidecar", "BlobStorageSidecar" or something others. "Sidecar" is too abstract here.

File: pkg/agent/sidecar/service/option.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+		if o.dirs != nil {
+			o.dirs = append(o.dirs, dirs...)
+		} else {
+			o.dirs = dirs
+		}
+		return nil
+	}
+}
+
+func WithDir(dir string) Option {
```
Comment (rinx) [inline pkg/agent/sidecar/service/option.go:88]: same.

File: internal/file/watch/option.go (hunk 2 above)
Comment (kpango) [reply internal/file/watch/option.go:54]: hmm, ok

File: internal/file/watch/watch.go (hunk 3 above)
Comment (kpango) [reply internal/file/watch/watch.go:102]: yes

File: internal/file/watch/watch.go (hunk 4 above)
Comment (kpango) [reply internal/file/watch/watch.go:135]: fsnotify recommend to watch event and error in a different goroutine.
https://github.com/howeyc/fsnotify/issues/7

File: internal/config/sidecar.go (hunk 5 above)
Comment (kpango) [reply internal/config/sidecar.go:20]: I see, I prefer  AgentSidecar sounds good

File: internal/file/watch/watch.go (hunk 4 above)
Comment (rinx) [reply internal/file/watch/watch.go:135]: I read it but i couldn't found a recommendation about that.
i think it's just about "not to create a fsnotify Watcher and monitor channels in a same goroutine".

I'm considering about the race condition about `init()` func when the `w.w.Errors` and `w.w.Events` closed simultaneously.

Comment (kpango) [reply internal/file/watch/watch.go:135]: sorry they recommend this thing in README.md
```
Do I have to watch the Error and Event channels in a separate goroutine?

As of now, yes. Looking into making this single-thread friendly (see howeyc #7)
```
and I agree with init race condition, I'll fix race condition problem about it but I think it's okay to run in different goroutine that mentioned in howeyc/fsnotify#7

Comment (rinx) [reply internal/file/watch/watch.go:135]: as described in README.md of fsnotify, it is only needed "*a* separate goroutine".
we don't have to have two goroutines here.

Comment (kpango) [reply internal/file/watch/watch.go:135]: a separate goroutine doesn't mean using different goroutine?

**Pull request merged**